File: python/moneysocket/core/message/crypt.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php


import json
import re

from Crypto.Cipher import AES
from Crypto import Random

from moneysocket.core.message.message import MoneysocketMessage


class MoneysocketCrypt():

    @staticmethod
    def is_clear(msg_bytes):
        try:
            msg_text = msg_bytes.decode("utf8", errors="strict")
            msg_dict = json.loads(msg_text)
        except:
            return False
        return True

    @staticmethod
    def is_cyphertext(msg_bytes):
        return not MoneysocketCrypt.is_clear(msg_bytes)

    @staticmethod
    def pad(msg_bytes):
        print("msg byte len: %d" % len(msg_bytes))
        BLOCK_SIZE = 16
        return msg_bytes + b'\0' * (BLOCK_SIZE - len(msg_bytes) % BLOCK_SIZE)

    @staticmethod
    def unpad(msg_bytes):
        return msg_bytes[:-ord(msg_bytes[len(msg_bytes) - 1:])]

    @staticmethod
    def encrypt(msg_bytes, shared_seed):
        aes256_key = shared_seed.derive_aes256_key()
        raw = MoneysocketCrypt.pad(msg_bytes)
        iv = Random.new().read(AES.block_size)
        cipher = AES.new(aes256_key, AES.MODE_CBC, iv)
        return iv + cipher.encrypt(raw)

    @staticmethod
    def decrypt(msg_bytes, shared_seed):
        aes256_key = shared_seed.derive_aes256_key()
        iv = msg_bytes[:16]
        cipher = AES.new(aes256_key, AES.MODE_CBC, iv)
        decrypted = cipher.decrypt(msg_bytes[16:])
        return re.sub(b'\x00*$', b'', decrypted)

    @staticmethod
    def wire_encode(msg, shared_seed=None):
        msg_bytes = json.dumps(msg).encode("utf8")
        if msg.MUST_BE_CLEARTEXT:
            return msg_bytes
        if shared_seed is None:
            return msg_bytes
        return MoneysocketCrypt.encrypt(msg_bytes, shared_seed)

    @staticmethod
    def wire_decode(msg_bytes, shared_seed=None):
        is_cyphertext = MoneysocketCrypt.is_cyphertext(msg_bytes)
        if is_cyphertext and shared_seed is None:
            return None, "no seed to decrypt cyphertext"

        if is_cyphertext:
            cleartext = MoneysocketCrypt.decrypt(msg_bytes, shared_seed)
            try:
                msg_text = cleartext.decode("utf8", errors="strict")
            except:
                return None, "message did not decode to utf8"

            msg, err = MoneysocketMessage.from_text(msg_text)
            if err:
                return None, err
            if msg.MUST_BE_CLEARTEXT:
                return None, "got encrypted msg which should have been clear"
            return msg, None
        else:
            try:
                msg_text = msg_bytes.decode("utf8", errors="strict")
            except:
                return None, "message did not decode to utf8"

            return MoneysocketMessage.from_text(msg_text)

File: browser/src/js/moneysocket/core/message/crypt.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

var aesjs = require("aes-js");
const Crypto = require('crypto');

const BinUtl = require('../../utl/bin.js').BinUtl;
const StringUtl = require('../../utl/string.js').StringUtl;

const MessageReceiver = require("./receiver.js").MessageReceiver;

class MoneysocketCrypt {

    static isClear(msg_bytes) {
        //console.log("checking clear: " + msg_bytes);
        try {
            var msg_txt = StringUtl.fromUtf8(msg_bytes);
            var msg_dict = JSON.parse(msg_txt);
        } catch (err) {
            //console.log("err: " + err);
            return false;
        }
        return true;
    }
    static isCyphertext(msg_bytes) {
        return ! MoneysocketCrypt.isClear(msg_bytes);
    }

    static pad(msg_bytes) {
        var pad_bytes = 16 - (msg_bytes.length % 16)
        var pad = new Uint8Array(pad_bytes);
        return BinUtl.arrayConcat(msg_bytes, pad);
    }

    static unpad(msg_bytes) {
        var i = msg_bytes.length - 1;
        while (msg_bytes[i] == 0) {
            i--;
        }
        return msg_bytes.slice(0, i + 1);
    }


    static encrypt(msg_bytes, shared_seed) {
        msg_bytes = MoneysocketCrypt.pad(msg_bytes);
        var key = shared_seed.deriveAes256Key();
        var iv = Crypto.randomBytes(16);
        var cbc = new aesjs.ModeOfOperation.cbc(key, iv);
        var encrypted = cbc.encrypt(msg_bytes);
        return BinUtl.arrayConcat(iv, encrypted);
    }

    static decrypt(msg_bytes, shared_seed) {
        var key = shared_seed.deriveAes256Key();
        var iv = msg_bytes.slice(0, 16);
        var cbc = new aesjs.ModeOfOperation.cbc(key, iv);
        var decrypted = cbc.decrypt(msg_bytes.slice(16));
        var unpadded = MoneysocketCrypt.unpad(decrypted);
        return unpadded;
    }

    static wireEncode(msg, shared_seed) {
        var msg_bytes = StringUtl.toUtf8(msg.toJson());
        if (msg.cryptLevel() == "CLEAR") {
            return msg_bytes;
        }
        if (shared_seed == null) {
            return msg_bytes;
        }
        return MoneysocketCrypt.encrypt(msg_bytes, shared_seed);
    }

    static wireDecode(msg_bytes, shared_seed) {
        //console.log("wire decode: " + msg_bytes);
        var is_cyphertext = MoneysocketCrypt.isCyphertext(msg_bytes);
        if (is_cyphertext && shared_seed == null) {
            return [null, "no seed to decrypt cyphertext"];
        }

        if (is_cyphertext) {
            var cleartext = MoneysocketCrypt.decrypt(msg_bytes, shared_seed);
            var msg_txt;
            try {
                msg_txt = StringUtl.fromUtf8(cleartext);
            } catch {
                return [null, "message did not decode to utf8"];
            }
            var [msg, err] = MessageReceiver.fromText(msg_txt);
            if (err != null) {
                console.log("fromtext err " + err);
                return [null, err];
            }
            if (msg.cryptLevel() == "CLEAR") {
                return [null, "got encrypted msg that should have been clear"];
            }
            return [msg, null];
        } else {
            var msg_txt;
            try {
                msg_txt = StringUtl.fromUtf8(msg_bytes);
            } catch {
                return [null, "message did not decode to utf8"];
            }
            var [msg, err] = MessageReceiver.fromText(msg_txt);
            if (err != null) {
                return [null, err];
            }
            if (msg.cryptLevel() != "CLEAR") {
                return [null, "got clear msg that should have been encrypted"];
            }
            return [msg, null];
        }
    }

}


exports.MoneysocketCrypt = MoneysocketCrypt;
